cluster/kubernetes/cached_disco.go

```diff
@@ -96,7 +96,32 @@ func makeCachedDiscovery(d discovery.DiscoveryInterface, c crd.Interface, shutdo
 	}
 
 	handler := handlerFn(cachedDisco)
-	store, controller := toolscache.NewInformer(lw, &crdv1beta1.CustomResourceDefinition{}, 5*time.Minute, handler)
+	store, controller := toolscache.NewInformer(lw, &crdv1beta1.CustomResourceDefinition{}, 0, handler)
+	go cachedDisco.invalidatePeriodically(shutdown)
 	go controller.Run(shutdown)
 	return cachedDisco, store, controller
 }
+
+func (d *cachedDiscovery) invalidatePeriodically(shutdown <-chan struct{}) {
+	// Make the first period shorter since we may be bootstrapping a
+	// newly-created cluster and the resource definitions may not be
+	// complete/stable yet.
+	initialPeriodDuration := 1 * time.Minute
+	subsequentPeriodsDuration := 5 * initialPeriodDuration
+	isInitialPeriod := true
```

Comment:
It's a minor stylistic thing, but I reckon this could be simpler if you use a timer:

```
delay := 1 * time.Minute
￼subsequentDelay := 5 * time.Minute
timer := time.NewTimer(delay)
￼for {
￼ 	select {
￼ 	case <-timer.C:
￼		timer.Reset(subsequentDelay)
￼ 		d.Invalidate()
￼ 	case <-shutdown:
￼ 		return
￼ 	}
￼}
```

Reply:
Thanks, that's much cleaner